## Импорт библиотек

In [105]:
import numpy as np
import pandas as pd

## Импорт данных

In [106]:

df_ef = pd.read_csv("https://raw.githubusercontent.com/EvgeniiSopovGitHub/datasets_uni/main/okaivd/lab_4_raw_data.csv")
display(df_ef.info())
display(df_ef)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 1 columns):
 #   Column                                                                                Non-Null Count  Dtype 
---  ------                                                                                --------------  ----- 
 0   date;unit_01;unit_02;unit_03;unit_04;unit_05;unit_06;unit_07;unit_08;unit_09;unit_10  365 non-null    object
dtypes: object(1)
memory usage: 3.0+ KB


None

date;unit_01;unit_02;unit_03;unit_04;unit_05;unit_06;unit_07;unit_08;unit_09;unit_10
0    2022-01-01;;35.356841405309794;29.501119790340...                                  
1    2022-01-02;37.1124746430136;49.94437515695568;...                                  
2    2022-01-03;;28.340224471077487;48.018449465161...                                  
3    2022-01-04;50.0;43.55556980883428;36.278831448...                                  
4    2022-01-05;42.434209274063136;28.7346745700960...                                  
..                                                 ...                                  
360  2022-12-27;40.82179355194795;;35.3529309426057...                                  
361  2022-12-28;36.92268068022749;42.95654078961374...                                  
362  2022-12-29;39.83574227370913;30.95874884595873...                                  
363  2022-12-30;34.484517731188724;36.8170607794108...                                  
364  2022-12-31;28.044284739156076;38.0537240632815...                                  

[365 rows x 1 columns]

<h3> Краткий вывод: </h3>
<ul>
  <li>Заголовки столбцов отображаются некоректно</li>
  <li>Строки присутствуют, подозрение, что данные представлены в некоретном виде</li>
  <li>Данные разделяются при помощи ";"</li>
</ul> 

In [107]:
# Исправление ошибки записи в csv файле
df_ef = pd.read_csv("https://raw.githubusercontent.com/EvgeniiSopovGitHub/datasets_uni/main/okaivd/lab_4_raw_data.csv"
                    , sep=";")
display(df_ef.info())
display(df_ef)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 11 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   date     365 non-null    object 
 1   unit_01  326 non-null    float64
 2   unit_02  320 non-null    float64
 3   unit_03  328 non-null    float64
 4   unit_04  330 non-null    float64
 5   unit_05  329 non-null    float64
 6   unit_06  321 non-null    float64
 7   unit_07  326 non-null    float64
 8   unit_08  324 non-null    float64
 9   unit_09  324 non-null    float64
 10  unit_10  340 non-null    float64
dtypes: float64(10), object(1)
memory usage: 31.5+ KB


None

date    unit_01    unit_02    unit_03    unit_04    unit_05  \
0    2022-01-01        NaN  35.356841  29.501120  49.620615  33.370093   
1    2022-01-02  37.112475  49.944375  42.270097  51.557284        NaN   
2    2022-01-03        NaN  28.340224  48.018449        NaN  39.209601   
3    2022-01-04  50.000000  43.555570  36.278831  43.076333        NaN   
4    2022-01-05  42.434209  28.734675  35.919586  49.739334  37.639153   
..          ...        ...        ...        ...        ...        ...   
360  2022-12-27  40.821794        NaN  35.352931  43.471269  40.233358   
361  2022-12-28  36.922681  42.956541  37.757330  39.423959        NaN   
362  2022-12-29  39.835742  30.958749  41.143042  46.283822        NaN   
363  2022-12-30  34.484518  36.817061  26.891212  40.649628  34.485905   
364  2022-12-31  28.044285  38.053724  50.317334  46.277700  44.742887   

       unit_06    unit_07    unit_08    unit_09    unit_10  
0    38.610154  34.559279  53.436849  41.238733  58.494774  
1    37.078793  22.265877  45.772941  75.733055  53.659106  
2    46.690520  25.886239  31.383527  78.607074        NaN  
3    35.764313  59.110809        NaN  39.244188  66.407508  
4          NaN  15.093752  50.664648  48.536941  65.719188  
..         ...        ...        ...        ...        ...  
360        NaN  25.770243  55.692111        NaN  68.647082  
361  42.500627  47.705860  30.529000  46.644630  60.888610  
362  35.878442  64.125878  52.910647  76.109909  65.518039  
363  39.078600  60.788903  48.904611  66.273623  67.848695  
364  35.568732  69.927246  56.475326  73.268611  63.756814  

[365 rows x 11 columns]

<h3> Краткий вывод: </h3>
<ul>
  <li>Заголовки столбцов отображаются коректно</li>
  <li>Некоторые строки имеют значение NaN</li>
  <li>Данные хорошо представлены</li>
</ul> 

## Заполнение пропусков

In [108]:
# Копирование и создание нового df для средних значений
df_ef_mean = df_ef.copy()

# заполнение NaN в столбцах с 1 по 11 средним значением
df_ef_mean[df_ef_mean.columns[1:11]] = df_ef_mean[df_ef_mean.columns[1:11]].fillna(df_ef_mean[df_ef_mean.columns[1:11]].mean())

display(df_ef_mean.info())
display(df_ef_mean)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 11 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   date     365 non-null    object 
 1   unit_01  365 non-null    float64
 2   unit_02  365 non-null    float64
 3   unit_03  365 non-null    float64
 4   unit_04  365 non-null    float64
 5   unit_05  365 non-null    float64
 6   unit_06  365 non-null    float64
 7   unit_07  365 non-null    float64
 8   unit_08  365 non-null    float64
 9   unit_09  365 non-null    float64
 10  unit_10  365 non-null    float64
dtypes: float64(10), object(1)
memory usage: 31.5+ KB


None

date    unit_01    unit_02    unit_03    unit_04    unit_05  \
0    2022-01-01  39.713383  35.356841  29.501120  49.620615  33.370093   
1    2022-01-02  37.112475  49.944375  42.270097  51.557284  40.549544   
2    2022-01-03  39.713383  28.340224  48.018449  39.977347  39.209601   
3    2022-01-04  50.000000  43.555570  36.278831  43.076333  40.549544   
4    2022-01-05  42.434209  28.734675  35.919586  49.739334  37.639153   
..          ...        ...        ...        ...        ...        ...   
360  2022-12-27  40.821794  39.598905  35.352931  43.471269  40.233358   
361  2022-12-28  36.922681  42.956541  37.757330  39.423959  40.549544   
362  2022-12-29  39.835742  30.958749  41.143042  46.283822  40.549544   
363  2022-12-30  34.484518  36.817061  26.891212  40.649628  34.485905   
364  2022-12-31  28.044285  38.053724  50.317334  46.277700  44.742887   

       unit_06    unit_07    unit_08    unit_09    unit_10  
0    38.610154  34.559279  53.436849  41.238733  58.494774  
1    37.078793  22.265877  45.772941  75.733055  53.659106  
2    46.690520  25.886239  31.383527  78.607074  60.231050  
3    35.764313  59.110809  49.864199  39.244188  66.407508  
4    39.931516  15.093752  50.664648  48.536941  65.719188  
..         ...        ...        ...        ...        ...  
360  39.931516  25.770243  55.692111  55.345836  68.647082  
361  42.500627  47.705860  30.529000  46.644630  60.888610  
362  35.878442  64.125878  52.910647  76.109909  65.518039  
363  39.078600  60.788903  48.904611  66.273623  67.848695  
364  35.568732  69.927246  56.475326  73.268611  63.756814  

[365 rows x 11 columns]

<h3> Краткий вывод: </h3>

<ul>
  <li>были заполнены пропуски</li>
  <li>display(df_ef_mean) Показал, что пропуски были заполненны одиннаковыми (для столбца, в которо пропуск) данными. Как пример unit_1 и его значение в 0 строке = [39.713383] и его значение в строке 2 = [39.713383]</li>
</ul> 


## Работа с данными (добавление новых столбцов)

In [109]:
# Название новых столбцов
new_columns = [
    'min',
    'median',
    'max',
]
display(df_ef_mean)

#for i in range (3):
    #df_ef_mean[new_columns[i]] = df_ef_mean.iloc[:, 1:11].min(axis=1)
    #размечтался

# Добавление новых столбцов
df_ef_mean[new_columns[0]] = df_ef_mean.iloc[:, 1:11].min(axis=1)
df_ef_mean[new_columns[1]] = df_ef_mean.iloc[:, 1:11].median(axis=1)
df_ef_mean[new_columns[2]] = df_ef_mean.iloc[:, 1:11].max(axis=1)
df_ef_mean['mean/std'] = df_ef_mean.iloc[:,1:11].mean(axis=1).round(2).astype(str) + '/' + df_ef_mean.iloc[:,1:11].std(axis=1).round(2).astype(str)

# Добавление строки с данными о среднем значении по столбцам
average_annual = df_ef_mean[df_ef_mean.columns[1:-1]].mean()
average_annual['date'] = 'average_annual'
df_ef_mean.loc[len(df_ef_mean)] = average_annual


display(df_ef_mean)

date    unit_01    unit_02    unit_03    unit_04    unit_05  \
0    2022-01-01  39.713383  35.356841  29.501120  49.620615  33.370093   
1    2022-01-02  37.112475  49.944375  42.270097  51.557284  40.549544   
2    2022-01-03  39.713383  28.340224  48.018449  39.977347  39.209601   
3    2022-01-04  50.000000  43.555570  36.278831  43.076333  40.549544   
4    2022-01-05  42.434209  28.734675  35.919586  49.739334  37.639153   
..          ...        ...        ...        ...        ...        ...   
360  2022-12-27  40.821794  39.598905  35.352931  43.471269  40.233358   
361  2022-12-28  36.922681  42.956541  37.757330  39.423959  40.549544   
362  2022-12-29  39.835742  30.958749  41.143042  46.283822  40.549544   
363  2022-12-30  34.484518  36.817061  26.891212  40.649628  34.485905   
364  2022-12-31  28.044285  38.053724  50.317334  46.277700  44.742887   

       unit_06    unit_07    unit_08    unit_09    unit_10  
0    38.610154  34.559279  53.436849  41.238733  58.494774  
1    37.078793  22.265877  45.772941  75.733055  53.659106  
2    46.690520  25.886239  31.383527  78.607074  60.231050  
3    35.764313  59.110809  49.864199  39.244188  66.407508  
4    39.931516  15.093752  50.664648  48.536941  65.719188  
..         ...        ...        ...        ...        ...  
360  39.931516  25.770243  55.692111  55.345836  68.647082  
361  42.500627  47.705860  30.529000  46.644630  60.888610  
362  35.878442  64.125878  52.910647  76.109909  65.518039  
363  39.078600  60.788903  48.904611  66.273623  67.848695  
364  35.568732  69.927246  56.475326  73.268611  63.756814  

[365 rows x 11 columns]

date    unit_01    unit_02    unit_03    unit_04    unit_05  \
0        2022-01-01  39.713383  35.356841  29.501120  49.620615  33.370093   
1        2022-01-02  37.112475  49.944375  42.270097  51.557284  40.549544   
2        2022-01-03  39.713383  28.340224  48.018449  39.977347  39.209601   
3        2022-01-04  50.000000  43.555570  36.278831  43.076333  40.549544   
4        2022-01-05  42.434209  28.734675  35.919586  49.739334  37.639153   
..              ...        ...        ...        ...        ...        ...   
361      2022-12-28  36.922681  42.956541  37.757330  39.423959  40.549544   
362      2022-12-29  39.835742  30.958749  41.143042  46.283822  40.549544   
363      2022-12-30  34.484518  36.817061  26.891212  40.649628  34.485905   
364      2022-12-31  28.044285  38.053724  50.317334  46.277700  44.742887   
365  average_annual  39.713383  39.598905  40.309659  39.977347  40.549544   

       unit_06    unit_07    unit_08    unit_09    unit_10        min  \
0    38.610154  34.559279  53.436849  41.238733  58.494774  29.501120   
1    37.078793  22.265877  45.772941  75.733055  53.659106  22.265877   
2    46.690520  25.886239  31.383527  78.607074  60.231050  25.886239   
3    35.764313  59.110809  49.864199  39.244188  66.407508  35.764313   
4    39.931516  15.093752  50.664648  48.536941  65.719188  15.093752   
..         ...        ...        ...        ...        ...        ...   
361  42.500627  47.705860  30.529000  46.644630  60.888610  30.529000   
362  35.878442  64.125878  52.910647  76.109909  65.518039  30.958749   
363  39.078600  60.788903  48.904611  66.273623  67.848695  26.891212   
364  35.568732  69.927246  56.475326  73.268611  63.756814  28.044285   
365  39.931516  37.464036  49.864199  55.345836  60.231050  28.343891   

        median        max     mean/std  
0    39.161769  58.494774   41.39/9.46  
1    44.021519  75.733055  45.59/13.96  
2    39.845365  78.607074   43.81/15.9  
3    43.315951  66.407508  46.39/10.04  
4    41.182863  65.719188  41.44/13.73  
..         ...        ...          ...  
361  41.525086  60.888610   42.59/8.12  
362  43.713432  76.109909  49.33/14.81  
363  39.864114  67.848695  45.62/14.55  
364  48.297517  73.268611  50.64/15.06  
365  41.811193  65.371048          NaN  

[366 rows x 15 columns]

## Экспрот dataframe в формате CSV файла

In [111]:
df_ef_mean.to_csv("lab4_bsv_stepanov.csv", sep=',', index=False, encoding='utf-8')